### ISSUE TITLE
[MCHECKSTYLE-406] Make plugin compatible with Checkstyle 8.44 (Upgraded to 9.1)

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-checkstyle-plugin#318]
**[Alexander Kriegisch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kriegaex)** opened **[MCHECKSTYLE-406](https://issues.apache.org/jira/browse/MCHECKSTYLE-406?redirect=false)** and commented

On GitHub, Dependabot suggests upgrading Checkstyle to something more recent than 8.41.1, but the corresponding builds fail, both for 8.43 and 8.44, e.g. [this build](https://github.com/dev-aspectj/aspectj-maven-plugin/pull/20/checks?check_run_id=2933248023):

```
Error:  Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.9.1:site (default-site) on project aspectj-maven-plugin: Error generating maven-checkstyle-plugin:3.1.2:checkstyle report: Failed during checkstyle configuration: cannot initialize module TreeWalker - cannot initialize module JavadocMethod - Property 'scope' does not exist, please check the documentation -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.9.1:site (default-site) on project aspectj-maven-plugin: Error generating maven-checkstyle-plugin:3.1.2:checkstyle report
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    (...)
Caused by: org.apache.maven.plugin.MojoExecutionException: Error generating maven-checkstyle-plugin:3.1.2:checkstyle report
    at org.apache.maven.plugins.site.render.SiteMojo.execute (SiteMojo.java:151)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    (...)
Caused by: org.apache.maven.reporting.MavenReportException: Failed during checkstyle configuration
    at com.puppycrawl.tools.checkstyle.api.AutomaticBean.configure (AutomaticBean.java:201)
    at org.apache.maven.plugins.checkstyle.exec.DefaultCheckstyleExecutor.executeCheckstyle (DefaultCheckstyleExecutor.java:167)
    at org.apache.maven.plugins.checkstyle.AbstractCheckstyleReport.executeReport (AbstractCheckstyleReport.java:533)
    (...)
```

This is in no way urgent, but as Dependabot keeps bugging me with this for every new Checkstyle version, I thought I should mention it.

---

**Affects:** 3.1.2

**Issue Links:**
- [MSHARED-990](https://issues.apache.org/jira/browse/MSHARED-990) Checkstyle config: Swap scope and exclude scope to access modifier in the check JavadocMethodCheck.
 (_**"is blocked by"**_)
- [MCHECKSTYLE-404](https://issues.apache.org/jira/browse/MCHECKSTYLE-404) maven-checkstyle-plugin version 3.1.2 Old version of checkstyle detected. Consider updating to >= v8.30
 (_**"is duplicated by"**_)

**Comment by jira-importer** [comment]
**[Piotr Zygielo](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pzygielo)** commented

The CS-plugin is configured to use CS 8.41.1, which works fine.

aspectj-maven-plugin is overriding this dependency - IMO it shall also be responsible for providing proper, up to date, CS configuration.

 

 

As the project aspectj-maven-plugin is not using customised CS config location, but the one configured in its parent

> \<configLocation>config/maven_checks.xml\</configLocation> \

and a-m-p is not providing customised configuration but

> [DEBUG] The resource 'config/maven_checks.xml' was found as jar:file:.../m2/repository/org/apache/maven/shared/maven-shared-resources/3/maven-shared-resources-3.jar!/config/maven_checks.xml.

is used.

Sounds like https://issues.apache.org/jira/browse/MSHARED-990

**Comment by jira-importer** [comment]
**[Alexander Kriegisch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kriegaex)** commented

I know what the preconfigured version is, thanks. But OTOH, users are encouraged to upgrade the dependency right in the first paragraph of the [plugin homepage](https://maven.apache.org/plugins/maven-checkstyle-plugin/):

> This version of the plugin uses Checkstyle 8.29 by default and requires Java 8. But you can upgrade the version used at runtime.

There is a direct link to another page, describing [how to do that](https://maven.apache.org/plugins/maven-checkstyle-plugin/examples/upgrading-checkstyle.html).

Like I said, it is not urgent, I simply wanted to create a ticket for you in order to make you aware of the fact that it does not work with the out-of-the-box configuration and that it might be a good idea to adjust that for the next release. If you think that the issue should be resolved in another component outside the scope of this plugin, that is fine with me. I have no way of knowing. I did not put checkstyle into the project, I found it there already. AFAIK, I am using the latest versions of everything in my POM, according to Dependabot, so I am not sure what I should upgrade.

As for the Mojohaus parent plugin config, those path names are not doing anything in AspectJ Maven, because the files simply do not exist. They are not mandatory.

I am not trying to complain about anything or pressurise you in any way. You might have noticed that I did not classify the issue as a bug but as a dependency upgrde. I am just a user trying to contribute a little bit of feedback to the maintainers who so generously provide me with this wonderful plugin, hoping that the feedback is useful and a future product version may benefit from it.

**Comment by jira-importer** [comment]
**[Piotr Zygielo](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pzygielo)** commented

> As for the Mojohaus parent plugin config, those path names are not doing anything in AspectJ Maven, because the files simply do not exist. They are not mandatory.

As I've already said:

> [DEBUG] The resource 'config/maven_checks.xml' was found as jar:[file:.../m2/repository/org/apache/maven/shared/maven-shared-resources/3/maven-shared-resources-3.jar!/config/maven_checks.xml|file:///.../m2/repository/org/apache/maven/shared/maven-shared-resources/3/maven-shared-resources-3.jar!/config/maven_checks.xml].

It comes from the build of aspectj-maven-plugin that failing build was linked in original description. So I'm not sure what could _those path names are not doing anything_ mean, as they are clearly doing something and are used. Otherwise - what CS configuration do you think a-m-p uses then?

https://github.com/dev-aspectj/aspectj-maven-plugin/pull/20 links to CS RN, and the https://checkstyle.org/releasenotes.html#Release_8.42 clearly enumerates **Breaking backward compatibility.**

> I am using the latest versions of everything in my POM, according to Dependabot, so

It's not enough. I am not a bot, but according to me, a-m-p uses CS config that is not compatible with newer CS, and if a-m-p desires to update CS, CS's config shall be updated as well.

> I am not sure what I should upgrade.

Already said: _aspectj-maven-plugin is overriding this dependency - IMO it shall also be responsible for providing proper, up to date, CS configuration._

**Comment by jira-importer** [comment]
**[Alexander Kriegisch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kriegaex)** commented

Thanks for your advice. Of course, it is an AspectJ Maven build, but what I said is true anyway: We do not use any specific config file there. We also have no wish to do so. We transparently use the default config from maven-shared-resources, which you were friendly enough to link to. We simply rely on that default config to work for us, we do not have any special needs there. We just use the plugin and are fine with the default Checkstyle version. No complaints at all. Like I said, this is in no way urgent. I just reacted on the Dependabot PR, checking the build log and decided to report that the latest Checkstyle versions do not fit with the current plugin version when using the default config. Nowhere did I say that this is a bug or your fault or that it even needs to be fixed in this project, if you think that the fix might be better done in Maven Shared Resources. Calm down, please, there is no blame here. So there is no need to be so defensive either. You made no mistake. Let us please talk nicely, shall we? 🙂

I know nothing whatsoever about Checkstyle. I simply followed the guide for this plugin, which I linked to before, upgrading Checkstyle according to the description there. My expectation is that in a **future** version, just like with this one, it works with the default config taken from Maven Shared Resources when using a more recent Checkstyle version. I hope it is OK to ask you if you can talk to the maintainers of the Shared Resources, whenever you think is the appropriate time, and come to an agreement with them to upgrade Checkstyle and the default config there in sync, so that it simply works for users.

Again: This is not a bug report, because there is no bug. It is simply a request to take the breaking changes in Checkstyle into account when upgrading both this plugin and the shared resources next time. We are talking about a future version here, not about the current one. I can wait. AspectJ Maven does not use Checkstyle 8.42+, we use 8.41.1 and are fine for now. Only the Dependabot PR suggests the update, but the PR is unmerged.  So no harm was done at all. Thank you very much.

**Comment by jira-importer** [comment]
**[Alexander Kriegisch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kriegaex)** commented

When Shared Resources 4 is released, this problem will go away, at least for the case reported here, by configuring the plugin as follows:

```xml
<pluginManagement>
  <plugins>
    <plugin>
      <groupId>org.apache.maven.plugins</groupId>
      <artifactId>maven-checkstyle-plugin</artifactId>
      <version>3.1.2</version>
      <dependencies>
        <dependency>
          <groupId>com.puppycrawl.tools</groupId>
          <artifactId>checkstyle</artifactId>
          <version>9.0.1</version>
        </dependency>
          <dependency>
              <groupId>org.apache.maven.shared</groupId>
              <artifactId>maven-shared-resources</artifactId>
              <version>4</version>
          </dependency>
      </dependencies>
    </plugin>
  </plugins>
</pluginManagement>
```

I.e., this plugin could be upgraded to Shared Resources 4, as soon as it is available on Maven Central.

**Comment by jira-importer** [comment]
**[Alexander Kriegisch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kriegaex)** commented

Update: maven-shared-resources version 4 is now available on Central. See also MSHARED-990. Consequently, the plugin dependency could be upgraded and a maintenance release published.

**Comment by jira-importer** [comment]
**[Alexander Kriegisch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kriegaex)** commented

Thanks for addressing this, [Sylwester Lachiewicz](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slachiewicz). BTW, is there a specific reason for the GitHub CI builds [using JDK 16-ea and 17-ea](https://github.com/apache/maven-checkstyle-plugin/blob/8a619aa0f2384a60e9af6b95d0724bb7cccee90a/.github/workflows/maven.yml#L28) instead of stable releases? Do you need a separate issue for fixing that?

**Comment by jira-importer** [comment]
**[Sylwester Lachiewicz](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slachiewicz)** commented

only leftover, fixed - thx for the reminder.
